kafka-client/src/main/java/dev/responsive/kafka/internal/db/rs3/RS3KVTable.java

```diff
@@ -0,0 +1,139 @@
+
+public class RS3KVTable implements RemoteKVTable<WalEntry> {
+  private static final Logger LOG = LoggerFactory.getLogger(RS3KVTable.class);
+
+  private final String name;
+  private final UUID storeId;
+  private final RS3Client rs3Client;
+  private final PssPartitioner pssPartitioner;
+  private LssId lssId;
+  private Long fetchOffset = ResponsiveStoreRegistration.NO_COMMITTED_OFFSET;
+  private RS3KVFlushManager flushManager;
+
+  public RS3KVTable(
+      final String name,
+      final UUID storeId,
+      final RS3Client rs3Client,
+      final PssPartitioner pssPartitioner
+  ) {
+    this.name = Objects.requireNonNull(name);
+    this.storeId = Objects.requireNonNull(storeId);
+    this.rs3Client = Objects.requireNonNull(rs3Client);
+    this.pssPartitioner = Objects.requireNonNull(pssPartitioner);
+  }
+
+  @Override
+  public KVFlushManager init(final int kafkaPartition) {
+    if (flushManager != null) {
+      throw new IllegalStateException("already initialized");
+    }
+
+    this.lssId = new LssId(kafkaPartition);
+
+    // TODO: we should write an empty segment periodically to any PSS that we haven't
+    //       written to to bump the written offset
+    final HashMap<Integer, Optional<Long>> lastWrittenOffset = new HashMap<>();
+    for (final int pss : pssPartitioner.allPss()) {
+      final var offsets = rs3Client.getCurrentOffsets(storeId, lssId, pss);
+      lastWrittenOffset.put(pss, offsets.writtenOffset());
+    }
+    final var fetchOffsetOrMinusOne = lastWrittenOffset.values().stream()
+        .map(v -> v.orElse(-1L))
+        .min(Long::compare)
+        .orElse(-1L);
+    if (fetchOffsetOrMinusOne == -1) {
+      this.fetchOffset = ResponsiveStoreRegistration.NO_COMMITTED_OFFSET;
+    } else {
+      this.fetchOffset = fetchOffsetOrMinusOne;
+    }
+
+    LOG.info("restore rs3 kv table from offset {} for {}", fetchOffset, kafkaPartition);
+
+    flushManager = new RS3KVFlushManager(
+        storeId,
+        rs3Client,
+        lssId,
+        this,
+        lastWrittenOffset,
+        kafkaPartition,
+        pssPartitioner
+    );
+    return flushManager;
+  }
+
+  @Override
+  public byte[] get(final int kafkaPartition, final Bytes key, final long minValidTs) {
+    final int pssId = pssPartitioner.pss(key.get(), kafkaPartition);
+    return rs3Client.get(storeId, lssId, pssId, flushManager.writtenOffset(pssId), key.get())
+        .orElse(null);
+  }
+
+  @Override
+  public KeyValueIterator<Bytes, byte[]> range(
+      final int kafkaPartition,
+      final Bytes from,
+      final Bytes to,
+      final long minValidTs
+  ) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public KeyValueIterator<Bytes, byte[]> all(final int kafkaPartition, final long minValidTs) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public long approximateNumEntries(final int kafkaPartition) {
+    LOG.warn("approximateNumEntries not implemented for RS3");
+    return 0;
```

Comment:
since we're not throwing an exception let's return -1 so it's obvious the store isn't empty, the API just isn't implemented